### PR TITLE
Imds healthstore integration

### PR DESCRIPTION
**Pull request by hglkrijger**
- adds a heartbeat signal for IMDS which fires every minute, and reports failure after 3 minutes
- updates unit tests
- update agent version to 2.2.27

### CODE REVIEW

File: azurelinuxagent/common/protocol/imds.py [hunk 1]
```diff
@@ -264,3 +268,52 @@ def get_compute(self):
...
+        Determines whether the metadata instance api returns 200, and the response
+        is valid: compute should contain location, name, subscription id, and vm size
+        and network should contain mac address and private ip address.
+        :return: Tuple<is_healthy:bool, error_response:str>
+            is_healthy: True when validation succeeds, False otherwise
+            error_response: validation failure details to assist with debugging
+        """
+
+        # ensure we get a 200
+        resp = restutil.http_get(self.instance_url, headers=self._headers)
```
[the remark this answers is not in the document]
Comment (hglkrijger) [reply azurelinuxagent/common/protocol/imds.py]: We do, see line 238.

[the remark this answers is not in the document]
Comment (harijayms) [reply azurelinuxagent/common/protocol/imds.py]: Hmm .. this seems to be the general Agent UserAgent, was trying to see if the health queries can have a different one

File: azurelinuxagent/common/protocol/imds.py [hunk 2]
```diff
@@ -11,7 +11,7 @@
 from azurelinuxagent.common.utils.flexible_version import FlexibleVersion
 
 IMDS_ENDPOINT = '169.254.169.254'
-APIVERSION = '2017-12-01'
+APIVERSION = '2018-02-01'
```
Comment (boumenot) [inline azurelinuxagent/common/protocol/imds.py]: This version is not available in sovereign clouds nor was the previous version.

/cc @harijayms

File: azurelinuxagent/common/protocol/imds.py [hunk 3]
```diff
@@ -264,3 +268,52 @@ def get_compute(self):
...
+            self.check_field(json_data['compute'], 'name')
+            self.check_field(json_data['compute'], 'subscriptionId')
+            self.check_field(json_data['compute'], 'vmSize')
+
+            self.check_field(json_data, 'network')
+            self.check_field(json_data['network'], 'interface')
+            self.check_field(json_data['network']['interface'][0], 'macAddress')
+            self.check_field(json_data['network']['interface'][0], 'ipv4')
+            self.check_field(json_data['network']['interface'][0]['ipv4'], 'ipAddress')
+            self.check_field(json_data['network']['interface'][0]['ipv4']['ipAddress'][0], 'privateIpAddress')
```
Comment (boumenot) [inline azurelinuxagent/common/protocol/imds.py]: None of these fields require API version 2018-02-01.  Do you need to verify that API?

File: azurelinuxagent/common/version.py [hunk 4]
```diff
@@ -113,7 +113,7 @@ def get_distro():
 
 AGENT_NAME = "WALinuxAgent"
 AGENT_LONG_NAME = "Azure Linux Agent"
-AGENT_VERSION = '2.2.25.1'
+AGENT_VERSION = '2.2.27'
```
Comment (boumenot) [inline azurelinuxagent/common/version.py]: Thank you!

File: azurelinuxagent/ga/monitor.py [hunk 5]
```diff
@@ -242,11 +248,13 @@ def collect_and_send_events(self):
     def daemon(self):
         min_delta = min(MonitorHandler.TELEMETRY_HEARTBEAT_PERIOD,
                         MonitorHandler.EVENT_COLLECTION_PERIOD,
-                        MonitorHandler.HOST_PLUGIN_HEARTBEAT_PERIOD).seconds
+                        MonitorHandler.HOST_PLUGIN_HEARTBEAT_PERIOD,
+                        MonitorHandler.IMDS_HEARTBEAT_PERIOD).seconds
         while self.should_run:
             self.send_telemetry_heartbeat()
             self.collect_and_send_events()
```
Comment (boumenot) [inline azurelinuxagent/ga/monitor.py]: Please open an issue to re-visit this.  I feel like we should be managing these individual things as a whole, or providing better support for them.

File: tests/data/imds/unicode.json [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  "compute": {
+    "location": "wéstus",
+    "name": "héalth",
+    "offer": "UbuntuSérvér",
+    "osType": "Linux",
+    "placementGroupId": "",
+    "platformFaultDomain": "0",
+    "platformUpdateDomain": "0",
+    "publisher": "Canonical",
+    "resourceGroupName": "tésts",
```
Comment (boumenot) [inline tests/data/imds/unicode.json]: Those are some fancy **é**'s you go there.

File: tests/data/imds/unicode.json [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    "sku": "16.04-LTS",
+    "subscriptionId": "21b2dc34-bcé6-4é63-9449-d2a8d1c2339é",
+    "tags": "",
+    "version": "16.04.201805220",
+    "vmId": "é7fdbfc4-2déb-4a4é-8615-éa6aaf50162é",
```
Comment (boumenot) [inline tests/data/imds/unicode.json]: Why are they fancy in the GUID?

File: azurelinuxagent/common/protocol/imds.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hglkrijger) [reply azurelinuxagent/common/protocol/imds.py]: Understood, will update.

File: azurelinuxagent/common/protocol/imds.py (hunk 2 above)
Comment (hglkrijger) [reply azurelinuxagent/common/protocol/imds.py]: Hari asked me to use this version.

File: azurelinuxagent/common/protocol/imds.py (hunk 3 above)
Comment (hglkrijger) [reply azurelinuxagent/common/protocol/imds.py]: I don't think it's required, but Hari asked for it. Happy to change.

File: azurelinuxagent/common/version.py (hunk 4 above)
Comment (hglkrijger) [reply azurelinuxagent/common/version.py]: Sure.

File: azurelinuxagent/common/protocol/imds.py (hunk 2 above)
Comment (harijayms) [reply azurelinuxagent/common/protocol/imds.py]: I havent updated the docs, but this version should be available now in national clouds.

File: tests/data/imds/unicode.json (hunk 7 above)
Comment (hglkrijger) [reply tests/data/imds/unicode.json]: No special reason - just did a find/replace for the unit test

File: azurelinuxagent/common/protocol/imds.py (hunk 3 above)
Comment (harijayms) [reply azurelinuxagent/common/protocol/imds.py]: we dont need the new API version , if you want we can use the older one .. agree that these dont need the new version

File: azurelinuxagent/ga/monitor.py (hunk 5 above)
Comment (hglkrijger) [reply azurelinuxagent/ga/monitor.py]: Ok, but I'm not quite sure what you mean. Let's sync offline.

**Comment by harijayms** [comment]
Could we also have a test for invalid JSON and Invalid response, like compute exists but there is no data for Name for example

**Comment by hglkrijger** [comment]
@harijayms - already there. Invalid json, empty string and missing field are all tested.

File: azurelinuxagent/common/protocol/imds.py (hunk 3 above)
Comment (hglkrijger) [reply azurelinuxagent/common/protocol/imds.py]: Spoke with Chris, we will leave as is.

File: azurelinuxagent/ga/monitor.py (hunk 5 above)
Comment (hglkrijger) [reply azurelinuxagent/ga/monitor.py]: Agreed - opened #1202

**Comment by hglkrijger** [comment]
@jasonzio - hope this doesn't cause too much rebasing pain for you..

**Comment by jasonzio** [comment]
Dammit...

**Pull request merged**